source/h2_decoder.c

```diff
@@ -270,27 +308,115 @@ void aws_h2_decoder_set_logging_id(struct aws_h2_decoder *decoder, void *id) {
+        DECODER_LOGF(
+            ERROR,
+            decoder,
+            "HPACK integer decoding decoded more data than was available '%s'",
+            aws_error_debug_str(aws_last_error()));
+
+        return AWS_HPACK_DECODE_ERROR;
+    }
+    decoder->frame_in_progress.payload_len -= (uint32_t)decoded_len;
+
+    return status;
+}
+
 /***********************************************************************************************************************
  * State functions
  **********************************************************************************************************************/
 
-static void s_decoder_set_state(struct aws_h2_decoder *decoder, const struct decoder_state *state) {
-    DECODER_LOGF(TRACE, decoder, "Moving from state %s to %s", decoder->state.name, state->name);
+static void s_decoder_run_state(
+    struct aws_h2_decoder *decoder,
+    const struct decoder_state *state,
+    struct aws_byte_cursor *input) {
 
+    DECODER_LOGF(TRACE, decoder, "Moving from state %s to %s", decoder->state.name, state->name);
     decoder->scratch.len = 0;
-    decoder->state = *state;
+
+    /* Special case for 0 length frames, otherwise frames could sit in incomplete until more data arrives */
+    if (state->bytes_required == 0) {
+        state->fn(decoder, input);
+    } else {
+        decoder->state = *state;
+    }
+}
+
+static void s_decoder_run_frame_state(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    AWS_ASSERT(decoder->frame_in_progress.type <= AWS_H2_FRAME_T_UNKNOWN);
+    s_decoder_run_state(decoder, s_state_frames[decoder->frame_in_progress.type], input);
 }
 
 static void s_decoder_reset_state(struct aws_h2_decoder *decoder) {
-    s_decoder_set_state(decoder, &s_state_length);
+    decoder->scratch.len = 0;
+    decoder->state = s_state_header;
 
     DECODER_LOG(TRACE, decoder, "Resetting frame in progress");
     AWS_ZERO_STRUCT(decoder->frame_in_progress);
 }
 
-static int s_state_fn_length(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+/** Returns as much of the current frame's payload as possible, and updates payload_len */
+static struct aws_byte_cursor s_decoder_get_payload(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+
+    struct aws_byte_cursor result;
+
+    const uint32_t remaining_length = decoder->frame_in_progress.payload_len;
+    if (input->len < remaining_length) {
+        AWS_FATAL_ASSERT(input->len <= UINT32_MAX);
+        result = aws_byte_cursor_advance(input, input->len);
+        decoder->frame_in_progress.payload_len -= (uint32_t)input->len;
+    } else {
+        result = aws_byte_cursor_advance(input, remaining_length);
+        decoder->frame_in_progress.payload_len = 0;
+    }
+
+    return result;
+}
+
+static int s_state_fn_header(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
 
-    AWS_FATAL_ASSERT(input->len >= 3);
+    AWS_FATAL_ASSERT(input->len >= 9);
```

Comment:
A comment on why this is 9 would be great. I have no idea reading this.

Reply:
I copy/pasted ascii art from the RFC into the code to explain these numbers